[kafka_consumer] Make args optional in kafka_consumer.py

jeffwidman opened the pull request:
### What does this PR do?

The consumer_groups, topics, and partitions arguments are each now optional. Any omitted values will be fetched from Zookeeper. If a value is omitted, the parent value must still be it's expected type (typically a dict). The `consumer_groups` argument can only be omitted if the`monitor_unlisted_consumer_groups` flag is set to True. The topics/partitions do not require this flag as I thought that specifying it at that lower level is redundant. Most users will specify a few consumer groups and leave it at that. 

Modern Kafka consumers store their offset in Kafka rather than Zookeeper, but this PR does not add support for this. Instead, I will be submitting a second kafka consumer offset check that will support offsets stored in Kafka. Originally I tried implementing that as part of this check, but the YAML interface got a bit clunky because it'd require a user to specify whether to fetch the offset from Zookeeper or Kafka. And if a user didn't specify a consumer group but instead set `monitor_unlisted_consumer_groups: True`, then it was totally ambiguous whether to check Kafka or Zookeeper. 

Additionally, since modern Kafka consumers have no dependency on Zookeeper, it seems like the better thing to do is keep them as two separate checks--one for zookeeper offsets and one for kafka offsets. This way the new check doesn't even concern itself with talking to Zookeeper at all, and it can do a more optimized handling of the Kafka connection since it's used for both broker and consumer offsets.

jeffwidman commented:
Migrated from https://github.com/DataDog/dd-agent/pull/2880 where @remh had reviewed and approved all these changes other than asking for default to a whitelist rather than default to checking all topics/partitions.

jeffwidman commented:
After thinking about this more, I think the whitelist of what to monitor for consumer lag should only specify the consumer group names, and deprecate listing the topics/partitions.

The most common use case for consumer lag is knowing consumer lag for all topics/partitions that a consumer group is consuming. I can't think of a time I'd want to only monitor the lag for a specific partition in a consumer group. Otherwise I have to remember every time I add new partitions to a topic to extend my check.

So an interface like:
    - kafka_connect_str: localhost:9092
      zk_connect_str: localhost:2181
      zk_prefix: /0.8
      consumer_groups:
        - my_consumer
        - cg2
      monitor_unlisted_consumer_groups: False

Pros:
 - This interface makes the check a lot cleaner both in configuration and in implementation.
 - It's easy to keep it backward compatible. 
 - Cleanly solves the problem of when a consumer group is subscribed to a regex pattern, or to topics that don't currently exist on the cluster but may later be created.
 - Potentially when we move to supporting kafka-based offsets it's basically the same number of network calls to get offsets for all topic_partitions consumed by the consumer group vs specific ones. I haven't implemented so I can't guarantee this until I do.
 - Don't have problems with users getting confused because they are trying to monitor topics/partitions that the consumer group isn't subscribed to


Cons:
 - Will output more data by default because it grabs all topic_partitions in the consumer group. 

I don't see this as a big deal because I suspect most users already are manually specifying all topic_partitions consumed by their consumer group. 

@remh unless you object, this is what I will do.

jeffwidman commented:
Added the `monitor_unlisted_consumer_groups` flag as requested by @remh along with a note about the downside of enabling it. So hopefully folks are incentivized to consider carefully before flooding your DB with Kafka consumer offset metrics. 

I am also working on a second kafka consumer offset check that will support offsets stored in Kafka. 
Originally I had that as part of this check, but the YAML interface got a bit clunky because it'd require a user to specify whether to fetch the offset from Zookeeper or Kafka. 

Since modern Kafka consumers have no dependency on Zookeeper, it seems like the better thing to do is keep them as two separate checks--one for zookeeper offsets and one for kafka offsets. This way the new check doesn't even concern itself with talking to Zookeeper at all, and it can do a more optimized handling of the Kafka connection since it's used for both broker and consumer offsets.

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -19,14 +24,11 @@ class KafkaCheck(AgentCheck):
-    instrument your individual kafka consumers to submit their offsets to
-    Datadog.
+    WARNING: Modern Kafka consumer store their offsets in Kafka rather than
+    zookeeper. You can monitor those offsets using the < TODO: TO BE NAMED > check.
```
We'll need to update this with the name of the new check whenever that gets figured out

jeffwidman commented:
@remh can you take a look at this? 

This is what we discussed at the Datadog summit in SF for checking offsets from zookeeper.

It's probably easiest reviewed as separate commits.

truthbk commented on kafka_consumer/check.py:
```diff
@@ -68,61 +205,99 @@ def check(self, instance):
-            # Get the broker offset
-            broker_offset = broker_offsets.get((topic, partition))
+        # Report the broker highwater offset
+        for (topic, partition), highwater_offset in highwater_offsets.iteritems():
```
Just a note. So there's a slight change in the logic here. I actually don't think this is a problem, as this would be a superset of broker highwatermark offsets comparing it to what the previous version of the check was doing (ie. every offset reported in the past will still be reported + more). The new version will essentially report all broker offsets collected in `_get_highwater_offsets()`, whereas before we only queried for broker offsets for which we had consumer offsets available. I don't think this would have any potential unwanted side-effects, but we should keep it in mind when testing.

jeffwidman replied:
This was a purposeful change. If the owner of a consumer group service comes to me complaining that they can't see data, I want to be able to diagnose from within datadog whether messages are making it into kafka as that guides whether I should focus on producer/consumer side. If broker highwater offsets are increasing, I know at least one producer is sending in messages.

truthbk commented on kafka_consumer/check.py:
```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+                        zk_path_partitions = zk_path_partition_tmpl % (consumer_group, topic)
+                        # Zookeeper returns the partition IDs as strings because
+                        # they are extracted from the node path
+                        partitions = [int(x) for x in self._get_zk_path_children(
```
Although we can be certain `partitions` here will have no repeated elements, we can't say the same about the list passed through by the user (in the YAML). So we should probably continue using a set to avoid unnecessary duplicated calls. Alternatively, we can maybe sanitize `consumer_groups` in `check()` before passing it on? Your call.

jeffwidman replied:
Good catch, I hadn't thought about user YAML. I'll defend against it somehow, personally I tend to error on being defensive / catching things early, so may implement both checks.

truthbk commented on kafka_consumer/conf.yaml.example:
```diff
@@ -18,6 +18,10 @@ instances:
+  #   # Setting monitor_unlisted_consumer_groups to True will tell the check to discover
+  #   # and fetch all offsets for all consumer groups stored in zookeeper. While this is
+  #   # often convenient, it can also put a lot of load on zookeeper, so use judiciously.
+  #   monitor_unlisted_consumer_groups: False
```
👍  I was wondering to myself if a feature flag was necessary... Then read this! It shows how it should definitely be behind a flag.

jeffwidman replied:
Additionally, from a business standpoint enabling this by default would cost Datadog a bit of money as it'd substantially increase the number of metrics emitted by most clusters. I prefer you guys keep prices low, so another good reason to disable by default.

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+                                "happen if the topic is currently being deleted.",
+                                topic, partition)
+                        elif error_code == 6:
+                            self.log.warn("Kafka broker returned "
```
Leader change should be uncommon for most clusters, as typically only happens when a broker is joing/leaving the cluster. Even then, just adding a new broker to a cluster can't trigger this state, it's only when the broker is part of the ISR list of some partition.

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+            topic_partitions_without_a_leader = []  # contains (topic, partition) tuples
+            for broker in kafka_client.cluster.brokers():
+                if not kafka_client.ready(broker.nodeId):
+                    time.sleep(0.5)
```
Yes, this is definitely a hack that I'd rather do differently. 

I also looked at caching the client instance, as well as trying to all the requests asynchronously, but ran into some design tradeoffs. You can see some of my notes on this in the source of #423 where caching `KafkaClient` has many more benefits.

1. The biggest drawback is that there's no way that I could find to trigger connection cleanup when the agent is stopped. Kafka added idle connection reaping in the brokers starting with 0.9. So I'd probably be fine with assuming in #423 that the broker is >= 0.9 as the only broker version that has kafka-based offsets but not idle-connection reaping is 0.8.2. Although it's still a little sketch if someone reports, "I ran #423 but you crashed my production 0.8.2 cluster" 😄 . However, this particular check, which is only fetching zookeeper offsets, is far more likely to be run against old brokers, so I'm hesitant to do that unless there's a clean way to teardown that connection upon graceful shutdown. 

2. Doing it async / caching the connection also adds a fair degree of complexity. It can certainly be done, but I'd rather get a working version live first without it. It's easier to debug if someone hits issues that aren't cropping up in either of our test environments. After 6 months of letting this check sit in production release of the agent, then I'd swap over to async.

3. Re-initializing this every time guarantees we'll have fresh metadata. Otherwise, we may need to trigger a metadata refresh in some scenarios. `KafkaClient` should do this automatically, but when I read the source code, I had just enough uncertainty that it would automatically refresh when I wanted (typically certain error conditions) that I shied away from trusting it. Not opposed to it if you or someone else wants to figure that out or just take a risk that it's doing the right thing. And if it's not doing the right thing, then arguably upstream `kafka-python` should be patched rather than trying to solve it here.

4. Even with all this, ideally we wouldn't have to `sleep` to fake blocking behavior. It'd be simpler to just block until the metadata refresh either finishes or times out. Again, I looked at the source of `kafka-python` for a while, but couldn't figure out how to achieve that. I have a feeling it's possible and I just missed it.

So with all that swirling around, that's why I stayed with the dirty hack. Very open to suggestions.

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+        where that broker is the leader:
+        https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetAPI(AKAListOffset)
+        """
+        # We could switch this to a long-lived conneciton that's cached
```
typo: conneciton

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -68,61 +205,99 @@ def check(self, instance):
+        # check first may be outdated by the time we check the other. Better to
+        # check consumer offset before checking broker offset because worst case
+        # is that overstates consumer lag a little. Doing it the other way can
+        # understate consumer lag to the point of having negative consumer lag,
```
"understate" --> "may understate"

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -68,61 +205,99 @@ def check(self, instance):
+            self.gauge('kafka.consumer_offset', consumer_offset, tags=consumer_group_tags)
+            if (topic, partition) not in highwater_offsets:
+                if (topic, partition) not in topic_partitions_without_a_leader:
+                    self.log.exception("Consumer offsets exist for topic: {topic} "
```
Should be `log.warn` probably

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -68,61 +205,99 @@ def check(self, instance):
+                # This is a really bad scenario because new messages produced to
+                # the topic are never consumed by that particular consumer
+                # group. So still report the negative lag as a way of increasing
+                # visibility of the error.
```
I'd rather do this at `ERROR`, as the user probably doesn't realize that their kafka consumer is completely missing new messages produced to the cluster. It's really a bad scenario.

jeffwidman replied:
@truthbk as discussed, this should be converted to emit an event into Datadog, not just a log event. I don't know the interface for doing this, so perhaps you can pull this PR in and then add that functionality in a later PR?

jeffwidman commented:
Updated to address comments.

Also updated logging to use the `log('%s', arg)` format as it's more performant in some cases. I did re-evaluate the log levels, and you can see where I landed based on the severity of the issue. If you disagree, I am willing to change them.

The one major outstanding issue is how to deal with the `sleep` hack. Personally, unless there's an obvious better solution, I'd rather see this land as-is as I originally submitted this PR back last September, so I'm a bit weary of waiting for it to be merged. And then update in a later PR... I'm not worried about forgetting as we'll have to solve this for the check that fetches consumer offsets from kafka, but that part still has some work left in it.

jeffwidman commented:
@truthbk As discussed, I removed the final commit that converted this check over to the new `KafkaClient` and will submit that as a new PR. 

So this one now only adds the `monitor_unlisted_consumer_groups` flag, with a default of `False`.

jeffwidman commented on kafka_consumer/check.py:
```diff
@@ -29,100 +38,196 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+            if (topic, partition) not in highwater_offsets:
+                self.log.warn("Consumer offsets exist for topic: {topic} "
+                    "partition: {partition} but that topic partition doesn't "
+                    "actually exist in the cluster.".format(**locals()))
```
I don't think this needs to be emitted as a Datadog event. 

This scenaario will arise if someone deletes a Kafka topic, but doesn't delete the consumer offsets. That's not a problem unless the topic gets recreated, in which case the consumer will miss new messages to the topic. However, in that case, the event triggered in https://github.com/DataDog/integrations-core/pull/271/files#r122269952 will notify regarding that.

jeffwidman commented:
@truthbk any update on this? 

As discussed, this is ready to be merged because I removed the final commit (over to #482), so now this PR only adds the `monitor_unlisted_consumer_groups` flag, with a default of `False`.

jeffwidman commented:
bump @truthbk

truthbk commented:
Finally working on the kafka PRs @jeffwidman. This one will be definitely in the next release. Going to give this one last pass, hopefully merge today.

truthbk commented:
Closing in favor of #612, thanks @jeffwidman - your work is included there. 🙇
